### DISCUSSION TITLE
Is it possible to use cursors in sqlalchemy with portals instead of named server side cursors?

### DISCUSSION THREAD

**Discussion by alvassin** [sqlalchemy/sqlalchemy#10174] [Usage Questions]
I found that SQLAlchemy provides AsyncConnection.stream() method, which i can use to get portions of data from postgres in memory efficient way:

```python
import asyncio
from sqlalchemy import text
from sqlalchemy.ext.asyncio import create_async_engine

async def main():
    engine = create_async_engine('postgresql://pg:pg@127.0.0.1:5432/pg')
    async with engine.connect() as conn:
        query = text('SELECT generate_series(0, 100000)')
        # Uses server side cursor
        async with conn.stream(
            query, execution_options={'yield_per': 5, 'stream_results': True}
        ) as cursor:
            async for row in cursor:
                print(f"{row}")
                await asyncio.sleep(0.1)

asyncio.run(main())
```

Postgresql extended query protocol (which is used by asyncpg) provides ability to prefetch specific number of rows from portal (seems similar to server side cursors). It seems it should be better from performance point of view (please tell me if I am wrong).

Is it possible to use portal instead of server side cursor somehow in SQLAlchemy?

**Comment by zzzeek** [comment]
> It seems it should be better from performance point of view (please tell me if I am wrong).

I'm not familiar with this feature but from its description I dont see why that would be the case, using a server side cursor rows are still fetched in batches, using a batch size that [can be controlled](https://docs.sqlalchemy.org/en/20/core/connections.html#streaming-with-a-fixed-buffer-via-yield-per).

Overall I would be very careful not to assume a. server side cursors are "faster" than a plain cursor or b. PG feature X is "faster" than a server side cursor.     What's "faster" is dependent on a wide range of factors and the actual "what's faster" can only be determined by running real profile tests with the exact scenario in use (that is, number of rows, bytes transferred per row, network characteristics, datatypes in use, even OS / python version, Python overhead ORM/core, etc.)

In my experience, server side cursors dont really offer a speed benefit, just  a way to limit the growth of memory for the case when fetching a very large result without using paging.
